## Final Project Submission

Please fill out:
* Student name: George Kariuki
* Student pace: part time 
* Scheduled project review date/time: 
* Instructor name: George Kamundia
* Blog post URL:


# 1. Business Understanding

## Project Context

Wilson Airport is expanding its operations to include aircraft acquisition for commercial and private aviation services. However, before investing millions of dollars into purchasing aircraft, the airport's Board of Directors requires a comprehensive risk assessment based on historical safety data.

The primary objective of this project is to leverage historical aviation accident data (1962–2023) to identify **low-risk aircraft makes and models** that would be suitable for purchase. The findings of this analysis will directly influence Wilson Airport's procurement strategy and fleet composition.

As data scientists, our role is to provide **data-driven insights** that support **strategic decision-making**, reduce **investment risk**, and promote **operational safety**.

---

## Business Questions

The business questions we are tasked to address are:

1. **Which aircraft makes and models have the lowest accident and fatality rates?**  
   We aim to identify aircraft that statistically demonstrate better safety records compared to others.

2. **Are newer aircraft models statistically safer than older models?**  
   We seek to determine whether more recently manufactured aircraft provide a measurable safety advantage.

3. **Are there specific types or categories of aircraft that should be preferred or avoided based on safety history?**  
   We will assess whether specific categories (e.g., single-engine vs twin-engine, commercial vs private) show significant safety differences.

4. **(Additional Question)**:  
   **Does the phase of flight (takeoff, cruise, landing) have a strong correlation with the severity of accidents for certain aircraft models?**  
   This self-derived question will help further refine safety strategies based on operational phases.

---

## Stakeholders

The key stakeholders who will rely on our findings include:

- **Board of Directors, Wilson Airport**  
  (Require high-level recommendations to inform aircraft purchasing decisions.)

- **Procurement and Acquisition Team**  
  (Need specific lists of recommended aircraft makes/models.)

- **Operations and Safety Department**  
  (Interested in trends related to aircraft safety history to adjust operational protocols.)

- **Investors and Financial Partners**  
  (Seek assurance that fleet investments are made into safer, lower-risk assets.)

- **Insurance Providers**  
  (May adjust premiums based on evidence of risk levels associated with the selected aircraft.)

---

## Expectations and Deliverables

In addition to answering the specific business questions outlined, we are expected to:

-  Perform data quality checks, basic cleaning, and exploratory analysis.
-  Use **visualizations** (bar charts, line graphs, pie charts) to highlight important findings clearly and simply.
-  Maintain a **reproducible, clean Jupyter Notebook** with clear Markdown framing and idiomatic code.
-  Provide **three clear, actionable business recommendations**.
-  Summarize findings in a **non-technical presentation** for business stakeholders.
-  Develop an **interactive dashboard** to explore accident trends and severity distributions.

---

## Business Value

The ability to make **evidence-based fleet acquisition decisions** will:

-  Reduce financial risks associated with buying unsafe or unreliable aircraft.
-  Increase customer confidence and brand reputation for Wilson Airport's new operations.
-  Improve long-term operational safety and minimize insurance costs.

By transforming raw historical data into meaningful, strategic insights, we enable Wilson Airport to make smart, safe, and profitable business moves.

---

![Aviation Safety](https://upload.wikimedia.org/wikipedia/commons/thumb/f/f3/Aviation_safety.png/640px-Aviation_safety.png)

[Learn more about aviation safety statistics here.](https://www.ntsb.gov/investigations/data/Pages/AviationDataStats.aspx)


# 2. Data Understanding

In this section, we aim to build a strong understanding of the aviation accident dataset.  
Before any cleaning, transformation, or analysis can take place, it is critical to:

- Explore the **overall structure** of the data.
- Assess the **completeness and quality** of the data.
- Understand the **meaning and role** of key columns.
- Detect any **missing values**, **strange data types**, or **anomalies**.
- Identify **time coverage** and any potential **bias** in the dataset.

**The specific steps we will follow:**

1. **Import libraries** required for data manipulation and visualization.
2. **Load** the dataset into a pandas DataFrame
3. **Inspect** the dataset’s structure (columns, data types, non-null counts).
4. **Preview** sample records to understand typical entries.
5. **Summarize** the dataset with basic descriptive statistics.
6. **Check for missing values** across all columns.
7. **Investigate key features** such as:
   - Aircraft Make
   - Aircraft Model
   - Event Date
   - Injury Severity
   - Aircraft Category
8. **Review time coverage**: What time range does the dataset cover?
9. **Identify data quality issues**, inconsistencies, or potential cleaning needs.
10. **Document** initial findings that will inform data cleaning and preparation steps later.

> Data understanding is a foundation of all strong data science projects. Skipping this phase risks introducing errors, misinterpretations, and poor business recommendations.


## Step 1: Import Required Libraries

To begin our data exploration, we need to import important Python libraries:

- **pandas**: for handling and manipulating dataframes.
- **numpy**: for numerical operations.
- **matplotlib** and **seaborn**: for creating visualizations to better understand patterns and trends in the data.


In [699]:
# Your code here - remember to use markdown cells for comments as well!
# Import essential libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set seaborn theme
sns.set_theme(style="whitegrid")


## Step 2: Load the Aviation Dataset

We will now load the aviation accident dataset into a pandas DataFrame for inspection.

Our goal in this step:
- Read the `.csv` file into memory.

## Step 3: Inspect the Dataset Structure

At this stage, we use the `.info()` method to inspect the **structure** of the aviation accident dataset.

**Purpose**:
- Understand the **number of records** and **columns**.
- Check **data types** of each feature (object, float, etc.).
- See **how many non-null entries** exist for each column (helps spot missing data early).

**Findings**:
- The dataset contains **90,348 records** and **31 columns**.
- Most columns are of **object** type (i.e., text or mixed data).
- Some columns (e.g., `Number.of.Engines`) are **numerical** (`float64`).
- Many important fields such as `Latitude`, `Longitude`, `Aircraft.Category`, and `FAR.Description` have **significant missing data**.
- Critical fields like `Event.Id`, `Accident.Number`, and `Event.Date` are mostly complete but **not 100% complete**.

**Next Step**:
- We will preview a few sample rows to get a better feel for what typical records look like.



In [700]:
# Load the aviation accident dataset
df = pd.read_csv('data/Aviation_Data.csv')

# Display basic information about the dataset
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90348 entries, 0 to 90347
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      90348 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50132 non-null  object 
 9   Airport.Name            52704 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87507 non-null  object 
 14  Make                    88826 non-null

C:\Users\HP\AppData\Local\Temp\ipykernel_15652\524183646.py:2: DtypeWarning: Columns (6,7,28) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('data/Aviation_Data.csv')


## Step 4: Preview Sample Records with `.head()`

The `.head()` method displays the **first five rows** of the dataset by default. This provides a quick look at typical entries and helps us understand:

* How the data is formatted (e.g., date formats, use of capital letters, empty cells).
* Common values in important columns like `Make`, `Model`, `Event.Date`, `Injury.Severity`, and `Aircraft.damage`.
* Which fields are frequently blank or contain placeholder values.

From the preview, we observed:

* Many records lack geographic data (`Latitude`, `Longitude`) and airport details.
* Common accident descriptors include values like `"Fatal(2)"` in `Injury.Severity`, `"Destroyed"` in `Aircraft.damage`, and `"Reciprocating"` in `Engine.Type`.
* Aircraft manufacturers like `Cessna`, `Piper`, and `Stinson` appear multiple times.
* Some columns, such as `FAR.Description` and `Air.carrier`, are often empty or unused.

This helps build early familiarity with the data and surfaces potential quality issues we'll address in later steps.

In [701]:
# Display the first five rows to understand the structure
df.head()


Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude  Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN        NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN        NaN          NaN   
2    Saltville, VA  United States  36.922223 -81.878056          NaN   
3       EUREKA, CA  United States        NaN        NaN          NaN   
4       Canton, OH  United States        NaN        NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

## Step 5: Summarize Dataset with Basic Descriptive Statistics (`.describe()`)

We generated basic descriptive statistics for the key numeric columns. Here are the main findings:

* **Number.of.Engines**
   * Mean: ~1.15 engines per aircraft.
   * Minimum: 0 (likely missing or error entries).
   * Maximum: 8 engines (likely large multi-engine aircraft).
* **Total.Fatal.Injuries**
   * Mean: ~0.65 fatalities per accident.
   * Standard Deviation: 5.49, indicating a wide spread — some accidents involve multiple fatalities.
   * Maximum: 349 fatalities (outlier, possibly large-scale accidents).
* **Total.Serious.Injuries**, **Total.Minor.Injuries**
   * Both have low means (~0.28 to 0.36), suggesting most accidents involve few or no serious/minor injuries.
* **Total.Uninjured**
   * Mean: ~5.32 uninjured individuals per accident.
   * High standard deviation (27.91), meaning in some cases, large groups of people were unharmed (e.g., commercial flights).

**Insights:**
* Most incidents involve small numbers of injuries or fatalities.
* Some large outliers (e.g., 349 fatalities) suggest commercial or mass-casualty incidents.
* Some accidents are recorded with 0 engines or 0 injuries, indicating data quality issues we'll need to handle later.

In [702]:
# Summarize numeric columns with descriptive statistics
df.describe()


Number.of.Engines  Total.Fatal.Injuries  Total.Serious.Injuries  \
count       82805.000000          77488.000000            76379.000000   
mean            1.146585              0.647855                0.279881   
std             0.446510              5.485960                1.544084   
min             0.000000              0.000000                0.000000   
25%             1.000000              0.000000                0.000000   
50%             1.000000              0.000000                0.000000   
75%             1.000000              0.000000                0.000000   
max             8.000000            349.000000              161.000000   

       Total.Minor.Injuries  Total.Uninjured  
count          76956.000000     82977.000000  
mean               0.357061         5.325440  
std                2.235625        27.913634  
min                0.000000         0.000000  
25%                0.000000         0.000000  
50%                0.000000         1.000000  
75%                0.000000         2.000000  
max              380.000000       699.000000

## Step 6: Check for Missing Values

We checked for missing values across all columns. Here are the main findings:

* Several columns have **substantial missing data**, particularly:
  * `Schedule` (~77,766 missing)
  * `Air.carrier` (~73,700 missing)
  * `FAR.Description` (~58,325 missing)
  * `Aircraft.Category` (~58,061 missing)
  * `Latitude` and `Longitude` (~55,000 missing each)
* Operational details like `Airport.Code`, `Airport.Name`, and `Broad.phase.of.flight` also have notable gaps.
* Critical outcome columns like `Total.Fatal.Injuries`, `Total.Serious.Injuries`, and `Total.Minor.Injuries` have around ~13,000 missing values.
* Core identifiers like `Event.Id`, `Event.Date`, and `Accident.Number` are missing for about ~1,459 records.

**Implications:**
* We may need to **drop** or **impute** certain columns depending on their importance for our analysis.
* Some columns with very high missingness (e.g., `Schedule`, `Air.carrier`) may not be reliable for modeling or analysis.

In [703]:
# Check for missing values across all columns
missing_values = df.isnull().sum().sort_values(ascending=False)

# Display columns with missing values
missing_values[missing_values > 0]


Schedule                  77766
Air.carrier               73700
FAR.Description           58325
Aircraft.Category         58061
Longitude                 55975
Latitude                  55966
Airport.Code              40216
Airport.Name              37644
Broad.phase.of.flight     28624
Publication.Date          16689
Total.Serious.Injuries    13969
Total.Minor.Injuries      13392
Total.Fatal.Injuries      12860
Engine.Type                8555
Report.Status              7843
Purpose.of.flight          7651
Number.of.Engines          7543
Total.Uninjured            7371
Weather.Condition          5951
Aircraft.damage            4653
Registration.Number        2841
Injury.Severity            2459
Country                    1685
Amateur.Built              1561
Model                      1551
Make                       1522
Location                   1511
Event.Date                 1459
Accident.Number            1459
Event.Id                   1459
dtype: int64

## Step 7: Investigate Key Features

We have explored the following key features to gather insights:

* **Aircraft Make**:
  * The top manufacturers are **Cessna** (22,227 occurrences) and **Piper** (12,029 occurrences), with a mix of capitalizations (e.g., `Cessna` vs. `CESSNA`, `Piper` vs. `PIPER`) indicating potential inconsistencies in the data.
* **Aircraft Model**:
  * **Cessna 152** (2,367 occurrences) and **Cessna 172** (1,756 occurrences) are the most common models. There are also a few rare models like "Rocket" and "ULTR," which might require further validation due to their low counts.
* **Event Date**:
  * The dataset spans from **October 24, 1948**, to **December 29, 2022**, providing a broad range of data over several decades.
* **Injury Severity**:
  * Most accidents were **Non-Fatal** (67,357 occurrences), with a smaller number of **Fatal** accidents. Notably, there are several variations of "Fatal" (e.g., **Fatal(1)**, **Fatal(2)**), which should be cleaned and standardized for better analysis.
* **Aircraft Category**:
  * The majority of accidents involve **Airplanes** (27,617 occurrences), followed by **Helicopters** (3,440 occurrences). There are also smaller numbers of **Gliders**, **Balloons**, and other aircraft types.

In [704]:
# Convert Event.Date to datetime
df['Event.Date'] = pd.to_datetime(df['Event.Date'], errors='coerce')

# Investigate important categorical features
print("Aircraft Make:\n", df['Make'].value_counts().head(10), "\n")
print("Aircraft Model:\n", df['Model'].value_counts().head(10), "\n")
print("Event Date Range:\n", df['Event.Date'].min(), "to", df['Event.Date'].max(), "\n")
print("Injury Severity:\n", df['Injury.Severity'].value_counts(), "\n")
print("Aircraft Category:\n", df['Aircraft.Category'].value_counts(), "\n")


Aircraft Make:
 Make
Cessna     22227
Piper      12029
CESSNA      4922
Beech       4330
PIPER       2841
Bell        2134
Boeing      1594
BOEING      1151
Grumman     1094
Mooney      1092
Name: count, dtype: int64 

Aircraft Model:
 Model
152          2367
172          1756
172N         1164
PA-28-140     932
150           829
172M          798
172P          689
182           659
180           622
150M          585
Name: count, dtype: int64 

Event Date Range:
 1948-10-24 00:00:00 to 2022-12-29 00:00:00 

Injury Severity:
 Injury.Severity
Non-Fatal     67357
Fatal(1)       6167
Fatal          5262
Fatal(2)       3711
Incident       2219
              ...  
Fatal(270)        1
Fatal(60)         1
Fatal(43)         1
Fatal(143)        1
Fatal(230)        1
Name: count, Length: 109, dtype: int64 

Aircraft Category:
 Aircraft.Category
Airplane             27617
Helicopter            3440
Glider                 508
Balloon                231
Gyrocraft              173
Weight-Shift      

## Step 8: Review Time Coverage

The dataset spans from **October 24, 1948**, to **December 29, 2022**, covering more than **74 years** of data. This broad time range gives us a comprehensive view of aviation accidents over several decades, including both historical and more recent incidents.

In [705]:
# Review time coverage of the dataset
print("Time range covered by Event.Date:", df['Event.Date'].min(), "to", df['Event.Date'].max())

Time range covered by Event.Date: 1948-10-24 00:00:00 to 2022-12-29 00:00:00


## Step 9: Identify Data Quality Issues

Here are the findings from the **data quality check**:

* **Duplicate Rows**:
  * There are **1,390 duplicate rows** in the dataset. These should be removed to ensure the analysis is accurate and not biased by repeated entries.
* **Columns with Missing Values**:
  * Several columns have high percentages of missing values, particularly:
     * **Latitude** (61.94% missing)
     * **Longitude** (61.95% missing)
     * **Aircraft.Category** (64.26% missing)
     * **FAR.Description** (64.56% missing)
     * **Schedule** (86.07% missing)
     * **Air.carrier** (81.57% missing)
  * These missing values may need to be imputed, dropped, or handled depending on the importance of these columns in answering the business questions.
* **Data Types**:
  * **Event.Date** has been correctly converted to `datetime64[ns]`, which will be useful for time series analysis.
  * Several columns such as **Latitude** and **Longitude** are still of type `object`. These should ideally be converted to numeric types for proper analysis.
  * Columns like **Number.of.Engines**, **Total.Fatal.Injuries**, **Total.Serious.Injuries**, etc., are already of type `float64`, which is appropriate for numerical analysis.

In [706]:
# Check for duplicate rows in the dataset
duplicates = df.duplicated().sum()
print(f"Number of duplicate rows: {duplicates}")

# Identify columns with high percentages of missing values
missing_values = df.isnull().mean() * 100
print(f"Columns with missing values percentage:\n{missing_values[missing_values > 50]}")

# Review data types to ensure they are correct
print("Data Types:\n", df.dtypes)


Number of duplicate rows: 1390
Columns with missing values percentage:
Latitude             61.944924
Longitude            61.954886
Aircraft.Category    64.263736
FAR.Description      64.555939
Schedule             86.073848
Air.carrier          81.573471
dtype: float64
Data Types:
 Event.Id                          object
Investigation.Type                object
Accident.Number                   object
Event.Date                datetime64[ns]
Location                          object
Country                           object
Latitude                          object
Longitude                         object
Airport.Code                      object
Airport.Name                      object
Injury.Severity                   object
Aircraft.damage                   object
Aircraft.Category                 object
Registration.Number               object
Make                              object
Model                             object
Amateur.Built                     object
Number.of.Engines 

# Data Preparation (Cleaning the Data)

## Remove Duplicate Rows

* Ensure each accident record is unique to prevent skewing results.
* Use:

    ```python
    df.drop_duplicates(keep='first', inplace=True)
    ```
* Track the number and percentage of duplicates removed.

## Handle Missing Data in Key Columns

* Focus only on columns essential to business insights:

    ###   Geographic Analysis

    * `Latitude`, `Longitude`: Create a binary column `Location.Available` and filter rows where both are available (i.e., both latitude and longitude are not missing).
    * `Country`, `Location`: Drop or impute if missing (not critical for analysis but helpful for geographical insights).

    ###   Aircraft Characteristics

    * `Make`, `Model`: Drop rows with missing values as these are key to understanding aircraft data.
    * `Aircraft.Category`: Drop rows with missing data since it’s important for category-based analysis.

    ###   Severity and Injury Analysis

    * `Injury.Severity`: Drop records with missing severity, as it's a key factor for any injury-related analysis.
    * `Total.Fatal.Injuries`, `Total.Serious.Injuries`, `Total.Minor.Injuries`, `Total.Uninjured`: Drop rows with missing values to ensure valid analysis of injury counts.

    ###   Operational Context

    * `Broad.phase.of.flight`: This column is not essential for core analysis, so it was excluded.
    * `Weather.Condition`: Drop rows with missing weather conditions to ensure the validity of weather-related analysis.

    ###   Aircraft Damage

    * `Aircraft.damage`: Drop rows with missing values as it’s crucial to correlate damage with injuries.

## Convert Data Types

* Convert `Latitude` and `Longitude` to `float64`, ensuring they are within valid geographic ranges.
* Ensure `Event.Date` and `Publication.Date` are in `datetime64[ns]` format.
* Convert injury and engine count fields to numeric types (either `float64` or `int64`).

## Standardize Categorical Values

* `Make`: Normalize case (e.g., "CESSNA", "Cessna" → "Cessna").
* `Injury.Severity`: Simplify into standard levels (e.g., "Fatal", "Non-Fatal", "Incident").
* `Aircraft.Category`: Normalize and group rare values as "Other".
* `Broad.phase.of.flight`: Group into broad categories (e.g., Takeoff, Cruise, Landing) if relevant.

## Validate Data Consistency

* Cross-check injury counts against `Injury.Severity`.
* Validate engine count against aircraft category.
* Ensure there are no contradictory values (e.g., "Fatal" with all injury counts = 0).

## Final Processing

* Reset index after all cleaning:

    ```python
    df.reset_index(drop=True, inplace=True)
    ```
* Output cleaned row count and summary statistics.
* Prepare the cleaned dataset for visualization and analysis.


## 1. Remove Duplicate Rows

**Objective:**  
Ensure data integrity by removing duplicate records, which can lead to bias in analysis and inaccurate results.

**Steps Taken:**

- Identified and removed duplicate rows from the dataset.
- Used the pandas function `df.drop_duplicates(keep='first')` to retain only the first occurrence of each unique record.
  
**Findings:**

- **Number of duplicates removed:** 0
- **Percentage of duplicates removed:** 0.00%

**Conclusion:**  
No duplicate records were found in the dataset, so no rows were removed. The dataset remains unchanged in terms of duplicate records.


In [707]:
# Remove duplicate rows
df = df.drop_duplicates(keep='first')

# Output the number of duplicates removed and the percentage
duplicates_removed = df.duplicated().sum()
total_records = len(df)
percentage_removed = (duplicates_removed / total_records) * 100

print(f"Number of duplicates removed: {duplicates_removed}")
print(f"Percentage of duplicates removed: {percentage_removed:.2f}%")



Number of duplicates removed: 0
Percentage of duplicates removed: 0.00%


# 2: Handle Missing Data

High-missingness columns (consider dropping if >70% missing):

* `Schedule` (86.07% missing)
* `Air.carrier` (81.57% missing)

These columns exhibit substantial missing data and might not provide significant value to the analysis. Dropping them could streamline the dataset.

In [708]:

# Columns with more than 70% missing
high_missing_cols = df.columns[df.isnull().mean() > 0.7]
print(f"Columns with more than 70% missing:\n{high_missing_cols}\n")

Columns with more than 70% missing:
Index(['Schedule', 'Air.carrier'], dtype='object')



In [709]:
# Drop columns with high missingness ( 'Schedule', 'Air.carrier'))
df = df.drop(columns=high_missing_cols)

### Handle Missing Data in Key Columns – Injury Severity

**Action Taken:**
- Dropped rows with missing `Injury.Severity`, which is essential for severity and injury analysis.
- **Rows dropped**: 1,000
- **New total rows**: 73,359

**Rationale:**
- The `Injury.Severity` column is critical for understanding the nature of accidents and their severity. Missing values here would significantly impact analysis related to injury severity classification and operational insights. Dropping rows with missing values ensures more accurate analysis. 


In [710]:
df['Injury.Severity'].isna().sum()


1069

In [711]:
# Store initial row count
initial_row_count = df.shape[0]

# Drop rows with missing Injury.Severity
df.dropna(subset=['Injury.Severity'], inplace=True)

# Report how many were dropped
rows_dropped = initial_row_count - df.shape[0]
print(f"Rows dropped due to missing Injury.Severity: {rows_dropped}")
print(f"New total rows: {df.shape[0]}")


Rows dropped due to missing Injury.Severity: 1069
New total rows: 87889


To address missing geographic data (Latitude and Longitude), we created a binary feature called `Location.Available`. This feature indicates whether both `Latitude` and `Longitude` values are available for a given record:

- **True**: Both `Latitude` and `Longitude` are non-null.
- **False**: Either `Latitude` or `Longitude` is missing.

This binary feature helps retain information about the presence or absence of geographic coordinates without discarding rows that may still be useful in analysis.




In [712]:
# For geographic data (Latitude and Longitude), consider creating a "location available" binary feature
df['Location.Available'] = df['Latitude'].notna() & df['Longitude'].notna()

In [713]:
# Check the distribution of the new "Location.Available" feature
print(df['Location.Available'].value_counts())

# Preview some records where location is available and unavailable
print(df[['Latitude', 'Longitude', 'Location.Available']].head())


Location.Available
False    53755
True     34134
Name: count, dtype: int64
    Latitude  Longitude  Location.Available
0        NaN        NaN               False
1        NaN        NaN               False
2  36.922223 -81.878056                True
3        NaN        NaN               False
4        NaN        NaN               False


In [714]:
# Drop rows where location data is not available
df = df[df['Location.Available']]

# Confirm the new shape
print(f"New total rows after dropping missing locations: {df.shape[0]}")


New total rows after dropping missing locations: 34134


This ensures your dataset only includes records with valid event dates, which are crucial for time-based analysis like trends over years or months.

In [715]:
# Drop rows with missing Event.Date
df.dropna(subset=['Event.Date'], inplace=True)

# Confirm rows were dropped
print(f"Remaining rows after dropping missing Event.Date: {df.shape[0]}")


Remaining rows after dropping missing Event.Date: 34134


rows were dropped in the critical injury columns (`Total.Fatal.Injuries`, `Total.Serious.Injuries`, `Total.Minor.Injuries`, and `Total.Uninjured`) after checking for missing values.

In [716]:
# Store the original row count before dropping rows
original_row_count = df.shape[0]

# Drop rows where any of the critical injury columns are missing
df.dropna(subset=['Total.Fatal.Injuries', 'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured'], inplace=True)

# Check how many rows were dropped
print(f"Rows dropped: {original_row_count - df.shape[0]} (Total rows before: {original_row_count})")



Rows dropped: 11742 (Total rows before: 34134)


### Handle Missing Data in Aircraft Damage

**Identify Missing Values:**

We first check for any missing values in the `Aircraft.damage` column using the `.isnull().sum()` method.
This helps us assess how many records have missing damage information.

**Drop Rows with Missing Data:**

We drop the rows where the `Aircraft.damage` column has missing values using the `.dropna(subset=['Aircraft.damage'])` method.
This ensures that we only keep records where damage information is available for analysis.

**Review Results:**

After dropping the missing data, we check the updated number of rows in the dataset to ensure the operation was successful.

In [717]:
# Check value counts in Aircraft.damage
print(df['Aircraft.damage'].value_counts(dropna=False))


Aircraft.damage
Substantial    18835
Destroyed       2406
NaN              585
Minor            502
Unknown           64
Name: count, dtype: int64


In [718]:
# Check for missing values in 'Aircraft.damage'
print(df['Aircraft.damage'].isnull().sum())

# Drop rows with missing 'Aircraft.damage'
df.dropna(subset=['Aircraft.damage'], inplace=True)

# Check the updated number of rows after dropping missing data
print(f"New total rows after dropping missing 'Aircraft.damage': {df.shape[0]}")


585
New total rows after dropping missing 'Aircraft.damage': 21807


### Handle Missing Values in Aircraft.Category

To ensure the integrity of the dataset, we removed rows where the `Aircraft.Category` column had missing (NaN) values. This step is crucial because `Aircraft.Category` is key to our analysis of different aircraft types and their relationship to accident severity and frequency.

**Action Taken:**

Used `df.dropna(subset=['Aircraft.Category'], inplace=True)` to remove rows where `Aircraft.Category` is missing.
The dataset after this operation contains 73,326 rows.

**Outcome:**

The rows with missing `Aircraft.Category` values have been removed, ensuring that our analysis reflects only complete records for this feature.

In [719]:
# Check unique values in 'Aircraft.Category' and their counts
print(df['Aircraft.Category'].value_counts())


Aircraft.Category
Airplane             18450
Helicopter            2315
Glider                 362
Weight-Shift           160
Gyrocraft              138
Balloon                 91
Powered Parachute       88
Ultralight              19
WSFT                     9
Unknown                  7
Rocket                   1
Blimp                    1
ULTR                     1
Name: count, dtype: int64


In [720]:
# Drop rows with missing values in 'Aircraft.Category'
df.dropna(subset=['Aircraft.Category'], inplace=True)

# Check the number of rows after dropping missing values
print(f"Rows after dropping missing 'Aircraft.Category': {df.shape[0]}")


Rows after dropping missing 'Aircraft.Category': 21642


**Drop Missing Values in the 'Make' Column**

In this step, we drop rows that have missing values in the `Make` column, as it's a key variable for analyzing aircraft characteristics.


In [721]:
# Count missing values in the 'Make' column
missing_make_count = df['Make'].isna().sum()

# Output the result
print(f"Missing values in 'Make' column: {missing_make_count}")


Missing values in 'Make' column: 2


In [722]:
# Drop rows with missing values in the 'Make' column
df.dropna(subset=['Make'], inplace=True)

# Output the new shape of the DataFrame after dropping rows
print(f"New total rows: {df.shape[0]}")


New total rows: 21640


### Drop Rows with Missing Weather Conditions

In this step, we focused on cleaning the data by removing rows with missing values in the `Weather.Condition` column. This ensures that weather data is available for all the remaining records, which is crucial for any analysis or visualization involving weather conditions.

In [723]:
# Check the unique values in the 'Weather.Condition' column
print(df['Weather.Condition'].value_counts())


Weather.Condition
VMC    19232
IMC      944
Unk      163
Name: count, dtype: int64


The `Weather.Condition` column was examined for unique values, which revealed:

* **VMC (Visual Meteorological Conditions):** 19,232 rows
* **IMC (Instrument Meteorological Conditions):** 944 rows
* **Unk (Unknown):** 163 rows

These values represent the meteorological conditions during the accident.

To ensure consistency in analysis:

The value 'Unk' may represent missing or unclear weather conditions. Depending on the analysis goals, these can be:

* **Treated as a separate category "Unknown"**
* **Or removed from the dataset if considered non-informative**

This step ensures that weather condition data is clean and categorized meaningfully for analysis and visualization.

In [724]:
# Drop rows with missing weather conditions
df.dropna(subset=['Weather.Condition'], inplace=True)

# Print the number of remaining rows
print(f"Remaining rows after dropping missing weather conditions: {df.shape[0]}")


Remaining rows after dropping missing weather conditions: 20339


## **Convert Data Types**

In this step, we ensure that each column is in the appropriate format for analysis:

1. **Latitude and Longitude:** 
   - These columns are converted from `object` to `float64` to facilitate geographic analysis. We also filter out any invalid coordinate values (outside the valid ranges for latitude and longitude).

2. **Event.Date and Publication.Date:**
   - These date columns are converted to `datetime64[ns]` to ensure that temporal analysis can be performed.

3. **Numerical Columns:** 
   - Columns like `Number.of.Engines` and injury counts are converted to `float64` or `int64` as needed for accurate statistical analysis.

**Outcome:**
- All columns are now in the appropriate data types for further analysis.


In [725]:

# Convert Latitude and Longitude to float64, setting invalid to NaN
df['Latitude'] = pd.to_numeric(df['Latitude'], errors='coerce')
df['Longitude'] = pd.to_numeric(df['Longitude'], errors='coerce')

# Create boolean masks for valid Latitude and Longitude ranges
valid_latitude = (df['Latitude'] >= -90) & (df['Latitude'] <= 90)
valid_longitude = (df['Longitude'] >= -180) & (df['Longitude'] <= 180)

# Optionally, flag invalid coordinates (retaining original rows)
df['Latitude_Invalid'] = ~valid_latitude
df['Longitude_Invalid'] = ~valid_longitude

# Convert Event.Date and Publication.Date to datetime64[ns], setting invalid to NaT
df['Event.Date'] = pd.to_datetime(df['Event.Date'], errors='coerce')
df['Publication.Date'] = pd.to_datetime(df['Publication.Date'], errors='coerce')

# Optionally, flag invalid dates (retaining original rows)
df['Event_Date_Invalid'] = df['Event.Date'].isna()
df['Publication_Date_Invalid'] = df['Publication.Date'].isna()

# Convert numerical columns to numeric, setting invalid to NaN
cols_to_numeric = ['Number.of.Engines', 'Total.Fatal.Injuries', 'Total.Serious.Injuries', 'Total.Minor.Injuries']
for col in cols_to_numeric:
    df[col] = pd.to_numeric(df[col], errors='coerce')

# Optionally, flag non-numeric values in these columns (retaining original rows)
for col in cols_to_numeric:
    df[f'{col}_NonNumeric'] = pd.isna(df[col]) & pd.notna(df[col].astype(str)) # Check if originally not NaN but became NaN after conversion

# Output the data types and a sample of the validity flags
print(df.dtypes)
print("\nOptional Validity Flags (First 5 rows):")
print(df[['Latitude_Invalid', 'Longitude_Invalid', 'Event_Date_Invalid', 'Publication_Date_Invalid'] + [f'{col}_NonNumeric' for col in cols_to_numeric]].head())

Event.Id                                     object
Investigation.Type                           object
Accident.Number                              object
Event.Date                           datetime64[ns]
Location                                     object
Country                                      object
Latitude                                    float64
Longitude                                   float64
Airport.Code                                 object
Airport.Name                                 object
Injury.Severity                              object
Aircraft.damage                              object
Aircraft.Category                            object
Registration.Number                          object
Make                                         object
Model                                        object
Amateur.Built                                object
Number.of.Engines                           float64
Engine.Type                                  object
FAR.Descript

C:\Users\HP\AppData\Local\Temp\ipykernel_15652\949084248.py:15: UserWarning: Parsing dates in %d-%m-%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df['Publication.Date'] = pd.to_datetime(df['Publication.Date'], errors='coerce')


### Standardize Categorical Variables
- **Action Taken**: The **Make** column was standardized by normalizing the case (e.g., "CESSNA" and "Cessna" were unified to "Cessna").
- **Result**: The **Make** column now contains consistently capitalized values for easier analysis.

In [726]:
# Standardize Aircraft Make by normalizing case
df['Make'] = df['Make'].str.strip().str.title()

# Check the unique values in 'Make' after standardization
print(df['Make'].value_counts().head(10))



Make
Cessna                 5177
Piper                  3066
Beech                  1098
Bell                    490
Mooney                  274
Robinson Helicopter     219
Air Tractor Inc         211
Robinson                211
Cirrus Design Corp      198
Boeing                  190
Name: count, dtype: int64


#### Standardize Categorical Variables (Injury Severity)
- **Action Taken**: The **Injury.Severity** column was standardized to group various fatal categories (e.g., 'Fatal(1)', 'Fatal(2)', etc.) into a single **Fatal** category, and **Non-Fatal** or **Incident** were grouped under **Non-Fatal**.
- **Result**: Injury severity is now simplified into two categories: **Fatal** and **Non-Fatal** for clearer analysis.



In [727]:
print(df['Injury.Severity'].value_counts())

Injury.Severity
Non-Fatal    16213
Fatal         3765
Minor          172
Serious        128
Fatal(1)        31
Fatal(2)        17
Fatal(3)         8
Incident         4
Fatal(5)         1
Name: count, dtype: int64


In [728]:
# Overwrite Injury.Severity with standardized values
df['Injury.Severity'] = df['Injury.Severity'].replace(
    to_replace=r'Fatal\(\d+\)', value='Fatal', regex=True
)

# Check updated value counts
print(df['Injury.Severity'].value_counts())


Injury.Severity
Non-Fatal    16213
Fatal         3822
Minor          172
Serious        128
Incident         4
Name: count, dtype: int64


### Standardize Categorical Values
### Broad.phase.of.flight:

The column contains multiple detailed flight phases. To simplify the analysis, these categories are grouped into broader phases:

* **Takeoff:** Includes "Takeoff", "Climb", "Go-around".
* **Landing:** Includes "Landing", "Approach", "Descent".
* **Cruise:** Includes "Cruise".
* **Other:** Includes "Maneuvering", "Taxi", "Standing".
* **Unknown:** Includes "Unknown".

This grouping helps reduce the number of unique values, making it easier to analyze and interpret flight phase data.

In [729]:
# List unique values in the 'Broad.phase.of.flight' column
print(df['Broad.phase.of.flight'].value_counts())


Broad.phase.of.flight
Cruise         23
Takeoff        19
Approach       17
Maneuvering    12
Landing         7
Descent         6
Climb           5
Unknown         4
Standing        1
Go-around       1
Name: count, dtype: int64


In [730]:
# Group the flight phases into broader categories
df['Broad.phase.of.flight'] = df['Broad.phase.of.flight'].replace({
    'Takeoff': 'Takeoff',
    'Climb': 'Takeoff',
    'Go-around': 'Takeoff',
    'Landing': 'Landing',
    'Approach': 'Landing',
    'Descent': 'Landing',
    'Cruise': 'Cruise',
    'Maneuvering': 'Other',  # Could be considered "Other" if it doesn't fit major categories
    'Taxi': 'Other',
    'Standing': 'Other',
    'Unknown': 'Unknown',
    'Other': 'Other'
})

# Check the updated values in Broad.phase.of.flight
print(df['Broad.phase.of.flight'].value_counts())


Broad.phase.of.flight
Landing    30
Takeoff    25
Cruise     23
Other      13
Unknown     4
Name: count, dtype: int64


### Verify no missing values in key columns

In [731]:
# List of key columns
key_columns = [
    'Location.Available',
    'Aircraft.Category',
    'Injury.Severity',
    'Aircraft.damage',
    'Make',
    'Weather.Condition',
    'Total.Fatal.Injuries',
    'Total.Serious.Injuries',
    'Total.Minor.Injuries',
    'Total.Uninjured',
]

# Check missing values in all key columns
missing_values = df[key_columns].isnull().sum()

# Display the missing values
print(missing_values)


Location.Available        0
Aircraft.Category         0
Injury.Severity           0
Aircraft.damage           0
Make                      0
Weather.Condition         0
Total.Fatal.Injuries      0
Total.Serious.Injuries    0
Total.Minor.Injuries      0
Total.Uninjured           0
dtype: int64


In [732]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 20339 entries, 50682 to 90345
Data columns (total 38 columns):
 #   Column                             Non-Null Count  Dtype         
---  ------                             --------------  -----         
 0   Event.Id                           20339 non-null  object        
 1   Investigation.Type                 20339 non-null  object        
 2   Accident.Number                    20339 non-null  object        
 3   Event.Date                         20339 non-null  datetime64[ns]
 4   Location                           20339 non-null  object        
 5   Country                            20339 non-null  object        
 6   Latitude                           100 non-null    float64       
 7   Longitude                          99 non-null     float64       
 8   Airport.Code                       14105 non-null  object        
 9   Airport.Name                       14329 non-null  object        
 10  Injury.Severity                    

### 🧹 Data Preparation Summary

- Removed duplicate records to ensure uniqueness.
- Filtered rows with valid geographic data (latitude & longitude).
- Dropped rows with missing values in key columns: aircraft details, injury severity, weather, and damage.
- Converted data types (dates, coordinates, injury counts) for consistency.
- Standardized categorical fields like `Make`, `Injury.Severity`, and `Aircraft.Category`.
- Final dataset cleaned and ready for analysis and visualization.
